hpx/lcos/future.hpp

```diff
@@ -413,126 +408,48 @@ namespace hpx { namespace lcos { namespace detail
 
-    template <typename Executor, typename Future, typename F>
-    inline typename hpx::traits::future_then_executor_result<
-        Executor, typename std::decay<Future>::type, F
-    >::type
-    then_execute_helper(Executor &&, F &&, Future &&);
+    // Dummy struct to call static assert with a false type
+    template <typename T>
+    struct dummy_false_type : std::false_type
+    {
+    };
 
     ///////////////////////////////////////////////////////////////////////////
     template <typename Future, typename FD, typename Enable = void>
-    struct future_then_dispatch;
-
-    // launch
-    template <typename Future, typename Policy>
-    struct future_then_dispatch<Future, Policy,
-        typename std::enable_if<
-            traits::is_launch_policy<Policy>::value
-        >::type>
+    struct future_then_dispatch
     {
-        template <typename Policy_, typename F>
-        HPX_FORCEINLINE
-        static typename hpx::traits::future_then_result<Future, F>::type
-        call(Future && fut, Policy_ && policy, F && f)
-        {
-            using result_type =
-                typename hpx::traits::future_then_result<Future, F>::result_type;
-            using continuation_result_type =
-                typename hpx::util::invoke_result<F, Future>::type;
-
-            typename hpx::traits::detail::shared_state_ptr<result_type>::type p =
-                detail::make_continuation_alloc<continuation_result_type>(
-                    hpx::util::internal_allocator<>{},
-                    std::move(fut), std::forward<Policy_>(policy),
-                    std::forward<F>(f));
-            return hpx::traits::future_access<future<result_type> >::create(
-                std::move(p));
-        }
 
-        template <typename Allocator, typename Policy_, typename F>
+        template <typename F>
         HPX_FORCEINLINE static
         typename hpx::traits::future_then_result<Future, F>::type
-        call_alloc(
-            Allocator const& alloc, Future&& fut, Policy_&& policy, F&& f)
+        call(Future&& fut, F&& f)
         {
-            using result_type =
-                typename hpx::traits::future_then_result<Future, F>::result_type;
-            using continuation_result_type =
-                typename hpx::util::invoke_result<F, Future>::type;
-
-            typename hpx::traits::detail::shared_state_ptr<result_type>::type p =
-                detail::make_continuation_alloc<continuation_result_type>(
-                    alloc, std::move(fut), std::forward<Policy_>(policy),
-                    std::forward<F>(f));
-            return hpx::traits::future_access<future<result_type> >::create(
-                std::move(p));
+            // dummy impl to fail compilation if this function is called
+            static_assert(dummy_false_type<Future>::value, "Cannot use the \
+                    dummy implementation of future_then_dispatch, please use \
+                    one of the template specialization.");
```

Comment:
Instead of the `dummy_false_type<Future>` you could simply say `sizeof(Future) == 0` (but that's not a necessary change). The error message may be improved however by referencing possibly missing #includes/module dependencies.